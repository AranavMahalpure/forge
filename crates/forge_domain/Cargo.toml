[package]
name = "forge_domain"
version = "0.1.0"
edition = "2024"

[dependencies]
derive_more = { version = "1.0.0-beta.6", features = ["from", "error", "display"] }
thiserror = "1.0.56"
url = "2.5.4"
