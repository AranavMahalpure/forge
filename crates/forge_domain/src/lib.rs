mod agent;
mod chat_request;
mod chat_response;
mod chat_stream_ext;
mod command;
mod config;
mod context;
mod conversation;
mod embedding;
mod environment;
mod errata;
mod error;
mod learning;
mod message;
mod model;
mod provider;
mod stream_ext;
mod tool;
mod tool_call;
mod tool_call_parser;
mod tool_call_service;
mod tool_choice;
mod tool_definition;
mod tool_name;
mod tool_result;
mod tool_usage;
mod user_interaction;

pub use agent::*;
pub use chat_request::*;
pub use chat_response::*;
pub use chat_stream_ext::*;
pub use command::*;
pub use config::*;
pub use context::*;
pub use conversation::*;
pub use embedding::*;
pub use environment::*;
pub use errata::*;
pub use error::*;
pub use learning::*;
pub use message::*;
pub use model::*;
pub use provider::*;
pub use stream_ext::*;
pub use tool::*;
pub use tool_call::*;
pub use tool_call_parser::*;
pub use tool_call_service::*;
pub use tool_choice::*;
pub use tool_definition::*;
pub use tool_name::*;
pub use tool_result::*;
pub use tool_usage::*;
pub use user_interaction::*;
