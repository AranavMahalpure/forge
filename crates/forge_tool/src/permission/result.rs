use std::path::Path;

use forge_domain::{PermissionError, PermissionResult};

/// Result display helper
#[derive(Debug)]
pub struct PermissionResultDisplay {
    allowed: bool,    
    message: String,
}

impl PermissionResultDisplay {
    /// Create simple display for permission result
    pub fn new(allowed: bool, path: &Path) -> Self {
        let message = format!(
            "Permission {} for {}",
            if allowed { "granted" } else { "denied" },
            path.display()
        );
        Self { allowed, message }
    }

    /// Create rich display with error message and permission details
    pub fn with_error(allowed: bool, path: &Path, error: Option<String>) -> Self {
        let mut message = format!(
            "Permission {} for {}",
            if allowed { "granted" } else { "denied" },
            path.display()
        );

        if let Some(err) = error {
            message = format!("{}: {}", message, err);
        }

        Self { allowed, message }
    }
}

impl std::fmt::Display for PermissionResultDisplay {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.message)
    }
}

pub trait AsPermissionDisplay {
    fn as_display(&self) -> PermissionResultDisplay;
}

impl<T> AsPermissionDisplay for PermissionResult<T> {
    fn as_display(&self) -> PermissionResultDisplay {
        match self {
            Ok(_) => PermissionResultDisplay::new(true, Path::new("/")),
            Err(PermissionError::OperationNotPermitted(msg)) => {
                PermissionResultDisplay::new(false, Path::new(msg))
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::path::PathBuf;

    #[test]
    fn test_display_simple() {
        let path = PathBuf::from("/test/path");
        let result = PermissionResultDisplay::new(true, &path);
        assert_eq!(result.to_string(), "Permission granted for /test/path");

        let result = PermissionResultDisplay::new(false, &path);
        assert_eq!(result.to_string(), "Permission denied for /test/path");
    }

    #[test]
    fn test_as_display() {
        let result: PermissionResult<()> = Ok(());
        let display = result.as_display();
        assert!(display.allowed);
        assert_eq!(display.to_string(), "Permission granted for /");

        let result: PermissionResult<()> =
            Err(PermissionError::OperationNotPermitted("test error".to_string()));
        let display = result.as_display();
        assert!(!display.allowed);
        assert_eq!(display.to_string(), "Permission denied for test error");
    }
}