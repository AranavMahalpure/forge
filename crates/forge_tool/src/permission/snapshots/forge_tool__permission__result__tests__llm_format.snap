---
source: crates/forge_tool/src/permission/result.rs
expression: display
---
<permission_result>
  <state>granted</state>
  <permission>Execute</permission>
  <path>/test/path</path>
  <context>User approved</context>
</permission_result>
